prov/sm2/src/sm2_progress.c

```diff
@@ -42,52 +42,187 @@
 #include "sm2.h"
 #include "sm2_fifo.h"
 
-static inline int sm2_cma_loop(pid_t pid, struct iovec *local,
-			       unsigned long local_cnt, struct iovec *remote,
-			       unsigned long remote_cnt, size_t total,
-			       bool write)
+static int sm2_progress_ipc_host_to_dev(struct sm2_ep *ep,
+					struct sm2_xfer_entry *xfer_entry,
+					struct fi_peer_rx_entry *rx_entry,
+					void *host_ptr)
 {
+	sm2_gid_t sender_gid = xfer_entry->hdr.sender_gid;
+	struct sm2_ipc_host_to_dev_data *host_to_dev_data;
+	struct ipc_info *ipc_info;
+	uint64_t device;
+	void *device_ptr = rx_entry->iov[0].iov_base;
+	struct ofi_mr **mr = (struct ofi_mr **) rx_entry->desc;
+	enum fi_hmem_iface iface = mr[0]->iface;
+	void *base;
+	int ret;
+
+	if (hmem_ops[iface].is_addr_valid(device_ptr, &device, NULL)) {
```

Comment:
Please add a similar ofi_is_addr_valid(iface...) call in ofi_hmem.h so we are consistent with how we call into the hmem ops array).

Reply:
not using this function anymore, as we discussed